### PR TITLE
feat(api): update consolidated multiple requests

### PR DESCRIPTION
**Pull request by Orta21**
Ref. metriport/metriport-internal#2090

Ticket: #2090 

### Dependencies

- Upstream: none
- Downstream: none

### Description

Be able to make multiple consolidated queries per unique param combo

### Testing

- Local
  - [x] Unit tests
  - [x] Send multiple consolidated queries
- Staging
  - [ ] Send multiple consolidated queries
- Production
  - [ ] Monitor

### Release Plan

- [x] Release packages
- [x] Update docs when flow is approved
- [ ] Merge this

### CODE REVIEW

File: packages/api/src/routes/medical/patient.ts
Comment (Orta21) [inline packages/api/src/routes/medical/patient.ts:289]: We are now sending an array of statuses so they can understand where each one is at

File: packages/api/src/command/medical/patient/consolidated-get.ts
Comment (Orta21) [inline packages/api/src/command/medical/patient/consolidated-get.ts:137]: Every time we run the consolidated query we want to clear out any progresses that were already completed or failed

File: packages/api/src/command/medical/patient/append-consolidated-query-progress.ts
Comment (Orta21) [inline packages/api/src/command/medical/patient/append-consolidated-query-progress.ts:22]: Wasnt being used

File: packages/api/src/command/medical/patient/__tests__/store-query-cmd.ts
Comment (jonahkaye) [inline packages/api/src/command/medical/patient/__tests__/store-query-cmd.ts:55]: Could we add a few more queries here so the tests show how this works when there are multiple queries?

File: packages/api/src/routes/medical/patient.ts
Comment (jonahkaye) [inline packages/api/src/routes/medical/patient.ts:283]: need to update the docstring for this function too

File: packages/api/src/routes/medical/patient.ts
Comment (jonahkaye) [inline packages/api/src/routes/medical/patient.ts:338]: update docstring here too

File: packages/api-sdk/src/medical/client/metriport.ts
Comment (jonahkaye) [inline packages/api-sdk/src/medical/client/metriport.ts:1]: no need to actually do fern deploy and follow the release process, as discussed [here](https://metriport.slack.com/archives/C04DMKE9DME/p1716571018692089), but you should either
1. update fern types here accordingly
2. add comments/TODOS to the fern types indicating that this change happened and when so when we do get around to updating fern types the next person knows to

File: packages/api/src/command/medical/patient/__tests__/store-query-cmd.ts
Comment (jonahkaye) [reply packages/api/src/command/medical/patient/__tests__/store-query-cmd.ts:55]: and also where there are querres where both are of conversion type pdf but have different other params

File: packages/api/src/command/medical/patient/__tests__/consolidated-get.test.ts
Comment (jonahkaye) [inline packages/api/src/command/medical/patient/__tests__/consolidated-get.test.ts:60]: can we add tests for when some of the request params are undefined to confirm the behavior of getCurrentConsolidatedProgress? 

Some of these have default behaviors that would make some of the checks in getCurrentConsolidatedProgress fail but would make two requests actually the same. 

For example, one request specifies conversion type of json. The other request doesnt. But json is the default. So they are the same actually but the `getCurrentConsolidatedProgress` wont catch that. Same thing for default of resources being all resources and specifying all resources being the same. 

So lets add tests for those cases and then modify the test function accordingly?

**Review by jonahkaye — CHANGES_REQUESTED** [review]
Actually had a few more comments. Sorry

File: packages/api/src/routes/medical/patient.ts
Comment (Orta21) [reply packages/api/src/routes/medical/patient.ts:283]: Wdym?

File: packages/api-sdk/src/medical/client/metriport.ts
Comment (Orta21) [reply packages/api-sdk/src/medical/client/metriport.ts:1]: I havent done this yet - is it updating the fern definitions? Like in `/fern/definition/medical/fhir.yml`?

File: packages/api/src/command/medical/patient/__tests__/store-query-cmd.ts
Comment (Orta21) [reply packages/api/src/command/medical/patient/__tests__/store-query-cmd.ts:55]: Create a function instead to create whatever permutation

File: packages/api/src/command/medical/patient/__tests__/consolidated-get.test.ts
Comment (Orta21) [reply packages/api/src/command/medical/patient/__tests__/consolidated-get.test.ts:61]: json i made as a default but added tests for the resources

File: packages/api-sdk/src/medical/client/metriport.ts
Comment (jonahkaye) [reply packages/api-sdk/src/medical/client/metriport.ts:1]: yes exactly

File: packages/api/src/command/medical/patient/consolidated-get.ts
Comment (jonahkaye) [inline packages/api/src/command/medical/patient/consolidated-get.ts:171]: 🎉

**Review by jonahkaye — APPROVED** [review]
lets add the fern thing but 🚢 conditional on that

File: packages/api/src/command/medical/patient/consolidated-get.ts
Comment (Orta21) [reply packages/api/src/command/medical/patient/consolidated-get.ts:136]: Yea to confirm its only to display the progress to cxs we dont use this for any webhook logic

File: packages/api/src/command/medical/patient/consolidated-get.ts
Comment (Orta21) [reply packages/api/src/command/medical/patient/consolidated-get.ts:82]: The idea is to send all the queries that are in progress as the return for this function. It seems odd that if there wasnt any that were in progress then we return all of them but if one was in progress we return one. 

What we could do is return a message along with it and say query with similar params is already in progress

File: packages/api/src/command/medical/patient/consolidated-get.ts
Comment (Orta21) [reply packages/api/src/command/medical/patient/consolidated-get.ts:164]: Yea good point

File: packages/api/src/command/medical/patient/consolidated-get.ts
Comment (leite08) [reply packages/api/src/command/medical/patient/consolidated-get.ts]: I think you're mixing two things:
- start consolidated query
- get consolidated query

What you referred to is the response of `GET /consolidated/query`, which I think it makes sense to be an array.

`startConsolidatedQuery` doesn't care about other consolidated queries, either there's a consolidated query w/ its parameters or not. If there is, it returns it.

And I think `POST /consolidated/query` should also always return a single consolidated query, I don't see a situation where it should return multiple.

File: packages/api-sdk/package.json
Comment (leite08) [inline packages/api-sdk/package.json:3]: The changes we're introducing are breaking changes, right? That should result in a major version bump.

File: packages/api/src/routes/medical/patient.ts
Comment (leite08) [inline packages/api/src/routes/medical/patient.ts:277]: This will break the API for current customers. I'd try to keep it backwards compatible. We can chat more on Slack if needed.

**Review by leite08 — CHANGES_REQUESTED** [review]

File: packages/api/src/command/medical/patient/__tests__/store-query-cmd.ts
Comment (Orta21) [reply packages/api/src/command/medical/patient/__tests__/store-query-cmd.ts:55]: Good point updated

File: packages/api/src/command/medical/patient/__tests__/consolidated-get.test.ts
Comment (Orta21) [reply packages/api/src/command/medical/patient/__tests__/consolidated-get.test.ts:118]: Is it fine to create a default at the top of the test in the beforeEach? And use that accordingly?

File: packages/api/src/routes/medical/patient.ts
Comment (Orta21) [reply packages/api/src/routes/medical/patient.ts:277]: Good point - to not introduce a breaking change immediately we have 2 options it seems

- Return status like we did before and add queries
- Create another endpoint with the new resp

I like the 2 option better lmk what you think

File: packages/api/src/command/medical/patient/__tests__/consolidated-get.test.ts
Comment (leite08) [reply packages/api/src/command/medical/patient/__tests__/consolidated-get.test.ts]: Its totally fine if we want a constant input. But some tests need a different input, so in those cases I think we should create it on the function itself.

WRT to using `beforeEach`:
- for regular stuff like creating a simple object, we can just declare the `const` at the start of the file, no need to use `beforeXxxx()`;
- `beforeEach` is useful when we want to run some code before each test - which means it might run 10s of times per file;
- that last point means we should try to use `beforeAll` instead of `beforeEach` when possible, so we minimize the cost of running tests - leading to faster tests, which helps with DX and CICD.

File: packages/api/src/routes/medical/patient.ts
Comment (leite08) [reply packages/api/src/routes/medical/patient.ts:277]: Since we're currently returning a JSON object, I think adding to it could work:
- `status` is the latest one;
- `queries` contain all the status.

Then we can mark `status` as deprecated and update the docs accordingly. We can communicate that to cxs and give them lots of time to update it to only rely on `queries`.

I like that because it doesn't break the behavior and cxs don't need to update their code to work with a new endpoint.

Thoughts?

File: docs/medical-api/api-reference/fhir/consolidated-data-query-post.mdx
Comment (leite08) [inline docs/medical-api/api-reference/fhir/consolidated-data-query-post.mdx:16]: nit: "...and you'll get the status and request ID of the previous one."

Thinking from the cx's perspective, it would be good to know whether a request to Metriport's API results in a new action or was it ignored and which one is running instead.

File: packages/api-sdk/src/medical/models/patient.ts
Comment (leite08) [inline packages/api-sdk/src/medical/models/patient.ts:29]: Can we add the `@deprecated` TSDoc to this prop here as well, please? When we add it to the type, anywhere we hover over that property we'll get this notification.

File: docs/medical-api/api-reference/fhir/consolidated-data-query-post.mdx
Comment (leite08) [inline docs/medical-api/api-reference/fhir/consolidated-data-query-post.mdx:84]: See `StartConsolidatedQueryProgressResponse`:
- It returns a JSON
- The JSON contains a single `query`, not an array

I'd also add a bit more context as to what the property `query` represents.

Thinking about a suggested text, I actually think we should return this:

```ts
type StartConsolidatedQueryProgressResponse = ConsolidatedQuery;
```

Why:
1. That's the actual representation of the initiated query, including the ID (`requestId`);
2. We don't use `message` AFAIK, so might as well remove it.

Then, my suggestion for explaining the response would be something like:

```
Returns the filters used for this query, in addition to the identifier used by the system to uniquely identify this query (property `requestId`).
```

And list the properties and a small text explaining what each represent.

File: fern/definition/medical/fhir.yml
Comment (leite08) [inline fern/definition/medical/fhir.yml:21]: update if we update the main docs

File: fern/definition/medical/fhir.yml
Comment (leite08) [inline fern/definition/medical/fhir.yml:26]: Isn't it missing the section explaining what it returns?

File: packages/api-sdk/package.json
Comment (leite08) [inline packages/api-sdk/package.json:3]: This is not `major`, this is `minor`. Major is the first number on semver.

File: packages/api-sdk/src/medical/models/fhir.ts
Comment (leite08) [inline packages/api-sdk/src/medical/models/fhir.ts:90]: Isn't this optional?

File: packages/api-sdk/src/medical/models/fhir.ts
Comment (leite08) [inline packages/api-sdk/src/medical/models/fhir.ts:104]: Do we need both `consolidatedFilterSchema` and `getConsolidatedFiltersSchema`?

File: packages/api-sdk/src/medical/models/fhir.ts
Comment (leite08) [inline packages/api-sdk/src/medical/models/fhir.ts:115]: Is `startedAt` ever not set? If not, could we make it required? Easier for users of the API if the prop is always present.

File: packages/api/src/command/medical/patient/__tests__/store-query-cmd.ts
Comment (leite08) [inline packages/api/src/command/medical/patient/__tests__/store-query-cmd.ts:57]: Just FYI: FakerJS has functions for this:
- [past](https://fakerjs.dev/api/date.html#past)
- [recent](https://fakerjs.dev/api/date.html#recent)

Tip: first define `to` and use it as a reference to `past` when building `from`, so its always before `to`/`recent`

File: packages/api/src/command/medical/patient/append-consolidated-query-progress.ts
Comment (leite08) [inline packages/api/src/command/medical/patient/append-consolidated-query-progress.ts:15]: nit: `Update a SINGLE patient's...`

File: packages/api/src/command/medical/patient/append-consolidated-query-progress.ts
Comment (leite08) [inline packages/api/src/command/medical/patient/append-consolidated-query-progress.ts:59]: This shouldn't happen, right? Maybe send a warning to Sentry?

File: packages/api/src/command/medical/patient/consolidated-get.ts
Comment (leite08) [inline packages/api/src/command/medical/patient/consolidated-get.ts:148]: I think we either rename this function to indicate we're only returning the current `processing`  progress, or remove the filter for `processing` and leave this to the client's to decide what to do with it.

I suggest the latter, so this function can be reused in other situations as well.

File: packages/api/src/command/medical/patient/consolidated-get.ts
Comment (leite08) [inline packages/api/src/command/medical/patient/consolidated-get.ts:173]: It seems we don't need this check, as we're already looking at the intersection of the arrays, no?

Comment (leite08) [reply packages/api/src/command/medical/patient/consolidated-get.ts:173]: Actually, using it introduces a bug when the length is the same but the contents are different (considering that my understanding is correct, that we only consider the same if the list of resources is also the same).

File: packages/api/src/command/medical/patient/consolidated-get.ts
Comment (leite08) [inline packages/api/src/command/medical/patient/consolidated-get.ts:190]: I don't understand why we need these.

This is saying that we consider the resources the same if:
- the existing query didn't have any resources but the new one is querying by all resources searchable by patient; or
- the new query is not filtering by patient but the previous one was querying by all resources searchable by patient.

File: packages/api/src/command/medical/patient/consolidated-get.ts
Comment (leite08) [inline packages/api/src/command/medical/patient/consolidated-get.ts:248]: nit: why checking `requestId`? This is already baked in `currentConsolidatedProgress`, right?

File: docs/_snippets/consolidated-query.mdx
Comment (leite08) [inline docs/_snippets/consolidated-query.mdx:2]: Maybe also mention that this is the same ID that's received as a result of `POST .../consolidated/query`?

File: fern/definition/medical/fhir.yml
Comment (Orta21) [reply fern/definition/medical/fhir.yml:26]: Isnt that in that in the response: docs: section?

File: packages/api-sdk/src/medical/models/fhir.ts
Comment (Orta21) [reply packages/api-sdk/src/medical/models/fhir.ts:104]: extended `consolidatedFilterSchema`

File: packages/api/src/command/medical/patient/__tests__/store-query-cmd.ts
Comment (Orta21) [reply packages/api/src/command/medical/patient/__tests__/store-query-cmd.ts:57]: Cool didnt know this

File: packages/api/src/command/medical/patient/consolidated-get.ts
Comment (Orta21) [reply packages/api/src/command/medical/patient/consolidated-get.ts:148]: Added param with `progressStatus = "processing"` as default

File: packages/api/src/command/medical/patient/consolidated-get.ts
Comment (Orta21) [reply packages/api/src/command/medical/patient/consolidated-get.ts:173]: We need the length because `const resp = getIsSameResources([], ["Observation"]);` will be true because 

queryResources = 0
intersection = 0

File: packages/api/src/command/medical/patient/consolidated-get.ts
Comment (Orta21) [reply packages/api/src/command/medical/patient/consolidated-get.ts:190]: Because empty array is the same as all the filters.

File: packages/api-sdk/src/medical/models/fhir.ts
Comment (Orta21) [reply packages/api-sdk/src/medical/models/fhir.ts:104]: Yea the dates are different

Comment (Orta21) [reply packages/api-sdk/src/medical/models/fhir.ts:104]: Ones string for filters we pass in and the other is dates

File: docs/_snippets/consolidated-query.mdx
Comment (Orta21) [reply docs/_snippets/consolidated-query.mdx:2]: This is being used on both POST and GET unless i remove this and add to both

File: packages/api-sdk/src/medical/models/fhir.ts
Comment (leite08) [reply packages/api-sdk/src/medical/models/fhir.ts]: NABD for this PR (we can address on a follow-up one):
- [bug] the SDK's `countPatientConsolidated()` returns `ConsolidatedCountResponse`, which has the dates as `Date`, but the endpoint returns those as `string`;
- [improvement] `getConsolidatedFiltersSchema` is extending `consolidatedFilterSchema`, but it overrides all properties - so no reason to extend it;
- [improvement] we probably can make `consolidatedCountSchema` use `getConsolidatedFiltersSchema` instead of `consolidatedFilterSchema`, it would simplify it
   - I'd look into using `Date` instead of `string`

File: packages/api/src/command/medical/patient/consolidated-get.ts
Comment (leite08) [reply packages/api/src/command/medical/patient/consolidated-get.ts:174]: I see the problem you reported, and I agree we need to account for that.

But what about returning `true` when the intersection is different but the lengths are the same?

For the issue you reported, we could check if `queryResources` empty or has zero items - then its not an intersection.

File: packages/api/src/command/medical/patient/consolidated-get.ts
Comment (leite08) [reply packages/api/src/command/medical/patient/consolidated-get.ts:191]: Gotcha, I see now. The naming was confusing me: `areQueryResourcesSearchableByPatient` doesn't give me the sense that it means "user included all resources searchable by patient".

File: docs/_snippets/consolidated-query.mdx
Comment (leite08) [reply docs/_snippets/consolidated-query.mdx:2]: The end result of the doc shouldn't be limited by how we structure our files. :)

I'll let you decide the best approach. 👍

**Review by leite08 — APPROVED** [review]
Approving so we can ship this, but I think we should prob look into these on a follow-up PR:
- bug with SDK/API mismatch - reported on [this comment](https://github.com/metriport/metriport/pull/2168#discussion_r1629974264).
- comparing resources - [comment](https://github.com/metriport/metriport/pull/2168#discussion_r1629981269)

**Pull request merged**
